### DISCUSSION TITLE
How to access frontmatter data (as title, author...) from a plugin ?

### DISCUSSION THREAD

**Discussion by Welsh44** [marp-team/marp#550] [Q&A] [answered]
Hi

I discovered Marp few weeks ago and found it very convenient.
I began with Windows binaries but went to npx marp-cli in order to use plugins.

As I 'm working on template for my company, I'd like to implement a canvas.
In order to this, I'd like to access via a plugin.
I found an example here : https://github.com/orgs/marp-team/discussions/510#discussioncomment-9397590
But I didn't find how to extend this code whith access to frontmatter "native" data as author, title, header, footer...

Thanks in advance for any clue.

**Comment by yhatt** [comment, marked as answer]
As mentioned at [https://marpit.marp.app/directives](https://marpit.marp.app/directives?id=custom-directives#:~:text=The%20returned%20key%2Dvalue%20will%20assign%20to%20marpitDirectives%20property%20in%20meta%20object%20of%20predetermined%20markdown%2Dit%20token(s)%20by%20the%20kind%20of%20directive.%20It%20would%20be%20useful%20for%20using%20assigned%20value%20in%20markdown%2Dit%20plugin.) and https://github.com/orgs/marp-team/discussions/510#discussioncomment-9397590, Marpit framework will store parsed directives, including defined in the front-matter, as the metadata of each slide token: `marpitDirectives`. You can inject [the parser rule of markdown-it](https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md) to read the metadata at after parsing directives.

```javascript
// engine.mjs
const yourPlugin = (md) => {
  md.core.ruler.after(
    'marpit_directives_apply',
    'your_custom_rule', // Rename the rule as you like
    (state) => {
      if (state.inlineMode) return

      for (const token of state.tokens) {
        if (
          typeof token.meta?.marpitSlide === 'number' &&
          typeof token.meta.marpitDirectives === 'object'
        ) {
          // Peak the metadata of directives for debug
          console.log(
            token.meta?.marpitSlide, // Slide page index (0-based)
            '=>',
            token.meta.marpitDirectives // Directives of the slide
          )

          // ... and you can do anything with using the value of directives ...
        }
      }
    }
  )
}

export default ({ marp }) => marp.use(yourPlugin)
```

```markdown
---
title: Title
author: Author
header: '**Header**'
footer: '_Footer_'
---

# Title

---

<!-- footer: Footer changed -->

Hello!
```

```shell
$ marp --engine ./engine.mjs ./markdown.md
[  INFO ] Converting 1 markdown...
0 => {
  header: '**Header**',
  footer: '_Footer_',
  marpCLITitle: 'Title',
  marpCLIAuthor: 'Author'
}
1 => {
  header: '**Header**',
  footer: 'Footer changed',
  marpCLITitle: 'Title',
  marpCLIAuthor: 'Author'
}
[  INFO ] markdown.md => markdown.html
```

> [!NOTE]
> The name of directive and the identifier of metadata may become different, for a reason of the purpose to avoid conflict identifiers with other plugins. Some of well-known global directives such as `author` and `title` are [the additional directives extended by Marp CLI](https://github.com/marp-team/marp-cli#metadata), so they have `marpCLI` prefixed identifiers. ([Code](https://github.com/marp-team/marp-cli/blob/main/src/engine/meta-plugin.ts))

**Comment by Welsh44** [comment]
Thank you.
I followed the links you mentionned but I didn't understand or wasn't aware of `token.meta?.marpitSlide`.
